File: Delivery_Time.R
# Input Variable = Sorting Time
# Output Variable = Delivery Time

#Import Dataset

Del.time <- read.csv(file.choose(), header = T)
colnames(Del.time) <- c("DT", "ST") # DT = Delivery Time & ST = Sorting Time
View(Del.time)
attach(Del.time)

# Exploratory data analysis
summary(Del.time)


install.packages("Hmisc")
library(Hmisc)
describe(Del.time) # To have a quick glance of a data


install.packages("lattice") # Highly used for data visualization
library("lattice") # dotplot is part of lattice package

# Graphical exploration
dotplot(Del.time$ST, main = "Dot Plot of Sorting Time")
dotplot(Del.time$DT, main = "Dot Plot of Delivery Time")

#Boxplot Representation

boxplot(Del.time$ST, col = "dodgerblue4")
boxplot(Del.time$DT, col = "red", horizontal = T)

#Histogram Representation

hist(Del.time$ST)
hist(Del.time$DT)

# Normal QQ plot
qqnorm(Del.time$ST)
qqline(Del.time$ST)

qqnorm(Del.time$DT)
qqline(Del.time$DT)


# Bivariate analysis
# Scatter plot
plot(Del.time$ST, Del.time$DT, main = "Scatter Plot", col = "Dodgerblue4", 
     col.main = "Dodgerblue4", col.lab = "Dodgerblue4", xlab = "Sorting Time", 
     ylab = "Delivery Time (Mins)", pch = 20)  # plot(x,y)


# Correlation Coefficient
cor(ST,DT)

# Linear Regression model
reg <- lm(DT ~ ST, data = Del.time) # Y ~ X
summary(reg)

confint(reg, level = 0.95)


pred <- predict(reg, interval = "predict")
pred <- as.data.frame(pred)

View(pred)


# ggplot for adding Regression line for data
library(ggplot2)

ggplot(data = Del.time, aes(ST, DT) ) +
  geom_point() + stat_smooth(method = lm, formula = y ~ x)


# Evaluation the model for fitness 
cor(pred$fit, Del.time$DT)

reg$residuals
rmse <- sqrt(mean(reg$residuals^2))
rmse


# Transformation Techniques
# transform the variables to check whether the predicted values are better

# input = log(x); output = y

plot(log(ST), DT)
cor(log(ST), DT)

reg_log <- lm(DT ~ log(ST), data = Del.time)
summary(reg_log)

confint(reg_log,level = 0.95)
pred <- predict(reg_log, interval = "predict")

pred <- as.data.frame(pred)
cor(pred$fit, Del.time$DT)

rmse <- sqrt(mean(reg_log$residuals^2))
rmse

# Regression line for data
ggplot(data = Del.time, aes(log(ST), DT) ) +
  geom_point() + stat_smooth(method = lm, formula = y ~ log(x))



# Log transformation applied on 'y'
# input = x; output = log(y)

plot(ST, log(DT))
cor(ST, log(DT))

reg_log1 <- lm(log(DT) ~ ST ,data= Del.time)
summary(reg_log1)

predlog <- predict(reg_log1, interval = "predict")
predlog <- as.data.frame(predlog)
reg_log1$residuals
sqrt(mean(reg_log1$residuals^2))

pred <- exp(predlog)  # Antilog = Exponential function
pred <- as.data.frame(pred)
cor(pred$fit, Del.time$DT)

res_log1 = DT - pred$fit
rmse <- sqrt(mean(res_log1^2))
rmse

# Regression line for data
ggplot(data = Del.time, aes(ST, log(DT)) ) +
  geom_point() + stat_smooth(method = lm, formula = log(y) ~ x)



# Non-linear models = Polynomial models
# input = x & x^2 (2-degree) and output = log(y)

reg2 <- lm(log(DT) ~ ST + I(ST*ST), data = Del.time)
summary(reg2)

predlog <- predict(reg2, interval = "predict")
pred <- exp(predlog)

pred <- as.data.frame(pred)
cor(pred$fit, Del.time$DT)

res2 = DT - pred$fit
rmse <- sqrt(mean(res2^2))
rmse

# Regression line for data
ggplot(data = Del.time, aes(ST, log(DT)) ) +
  geom_point() + stat_smooth(method = lm, formula = y ~ poly(x, 2, raw = TRUE))


# Data Partition

# Random Sampling
n <- nrow(Del.time)
n1 <- n * 0.8
n2 <- n - n1

train_ind <- sample(1:n, n1)
train <- Del.time[train_ind, ]
test <-  Del.time[-train_ind, ]


plot(log(train$ST), train$DT)
plot(log(test$ST), test$DT)

model <- lm(DT ~ log(ST), data = train)
summary(model)

confint(reg,level=0.95)

log_res <- predict(model,interval = "confidence", newdata = test)

predict_original <- as.data.frame(log_res)
test_error <- test$DT - predict_original$fit # calculate error/residual
test_error

test_rmse <- sqrt(mean(test_error^2))
test_rmse

log_res_train <- predict(reg_log,interval = "confidence", newdata = train)

predict_original_train <- as.data.frame(log_res_train)
train_error <- train$DT - predict_original_train$fit # calculate error/residual
train_error

train_rmse <- sqrt(mean(train_error^2))
train_rmse

File: Delivery_Time.py
# Importing necessary libraries
import pandas as pd # deals with data frame  
import numpy as np  # deals with numerical values

# Input Variable = Sorting Time
# Output Variable = Delivery Time

del_time = pd.read_csv("C:/Users/personal/Desktop/delivery_time.csv")
del_time.columns = "DT", "ST" # DT = Delivery Time & ST = Sorting Time
del_time.columns

# Exploratory data analysis:
del_time.describe()

#Graphical Representation
import matplotlib.pyplot as plt # mostly used for visualization purposes 

plt.bar(height = del_time.DT, x = np.arange(1,22, 1))
plt.hist(del_time.DT) #histogram
plt.boxplot(del_time.DT) #boxplot

plt.bar(height = del_time.ST, x = np.arange(1, 22, 1))
plt.hist(del_time.ST) #histogram
plt.boxplot(del_time.ST) #boxplot

# Scatter plot
plt.scatter(x = del_time['ST'], y = del_time['DT'], color = 'red') 

# correlation
np.corrcoef(del_time.ST, del_time.DT) 

# Covariance
cov_output = np.cov(del_time.ST, del_time.DT)[0, 1]
cov_output


# Import library
import statsmodels.formula.api as smf

# Simple Linear Regression
model = smf.ols('DT ~ ST', data = del_time).fit()
model.summary()

pred1 = model.predict(pd.DataFrame(del_time['ST']))

# Regression Line
plt.scatter(del_time.ST, del_time.DT)
plt.plot(del_time.ST, pred1, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()

# Error calculation
res1 = del_time.DT - pred1
res_sqr1 = res1 * res1
mse1 = np.mean(res_sqr1)
rmse1 = np.sqrt(mse1)
rmse1

######### Model building on Transformed Data
# Log Transformation
# x = log(Sorting Time); y = Delivery Time

plt.scatter(x = np.log(del_time['ST']), y = del_time['DT'], color = 'brown')
np.corrcoef(np.log(del_time.ST), del_time.DT) #correlation

model2 = smf.ols('DT ~ np.log(ST)', data = del_time).fit()
model2.summary()

pred2 = model2.predict(pd.DataFrame(del_time['ST']))

# Regression Line
plt.scatter(np.log(del_time.ST), del_time.DT)
plt.plot(np.log(del_time.ST), pred2, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()

# Error calculation
res2 = del_time.DT - pred2
res_sqr2 = res2 * res2
mse2 = np.mean(res_sqr2)
rmse2 = np.sqrt(mse2)
rmse2


#### Exponential transformation
# x = Sorting Time; y = log(Delivery Time)

plt.scatter(x = del_time['ST'], y = np.log(del_time['DT']), color = 'orange')
np.corrcoef(del_time.ST, np.log(del_time.DT)) #correlation

model3 = smf.ols('np.log(DT) ~ ST', data = del_time).fit()
model3.summary()

pred3 = model3.predict(pd.DataFrame(del_time['ST']))
pred3_at = np.exp(pred3)
pred3_at

# Regression Line
plt.scatter(del_time.ST, np.log(del_time.DT))
plt.plot(del_time.ST, pred3, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()

# Error calculation
res3 = del_time.DT - pred3_at
res_sqr3 = res3 * res3
mse3 = np.mean(res_sqr3)
rmse3 = np.sqrt(mse3)
rmse3


#### Polynomial transformation
# x = ST; x^2 = ST*ST; y = log(DT)

model4 = smf.ols('np.log(DT) ~ ST + I(ST*ST)', data = del_time).fit()
model4.summary()

pred4 = model4.predict(pd.DataFrame(del_time))
pred4_at = np.exp(pred4)
pred4_at

# Regression line
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 2)
X = del_time.iloc[:, 0:1].values
X_poly = poly_reg.fit_transform(X)


plt.scatter(del_time.ST, np.log(del_time.DT))
plt.plot(X, pred4, color = 'red')
plt.legend(['Predicted line', 'Observed data'])
plt.show()


# Error calculation
res4 = del_time.DT - pred4_at
res_sqr4 = res4 * res4
mse4 = np.mean(res_sqr4)
rmse4 = np.sqrt(mse4)
rmse4


# Choose the best model using RMSE
data = {"MODEL":pd.Series(["SLR", "Log model", "Exp model", "Poly model"]), "RMSE":pd.Series([rmse1, rmse2, rmse3, rmse4])}
table_rmse = pd.DataFrame(data)
table_rmse

###################
# The best model - From the RMSE Log Model 1 is the best model with Low Rmse

from sklearn.model_selection import train_test_split

train, test = train_test_split(del_time, test_size = 0.2)

finalmodel = smf.ols('DT ~ np.log(ST)', data = train).fit()
finalmodel.summary()


# Predict on test data
test_pred = finalmodel.predict(pd.DataFrame(test))
test_pred

# Model Evaluation on Test data
test_res = test.DT - test_pred
test_sqrs = test_res * test_res
test_mse = np.mean(test_sqrs)
test_rmse = np.sqrt(test_mse)
test_rmse


# Prediction on train data
train_pred = finalmodel.predict(pd.DataFrame(train))
train_pred

# Model Evaluation on train data
train_res = train.DT -  train_pred
train_sqrs = train_res * train_res
train_mse = np.mean(train_sqrs)
train_rmse = np.sqrt(train_mse)
train_rmse


